### PR TITLE
CFE-3006: Move known_dirs to libutils

### PR DESCRIPTION
**Pull request by karlhto**
Made some code style changes as well, as well as changing the order of the source files in Makefile.am (for libutils and libpromises) to alphabetical instead of seemingly arbitrary.

Ticket: CFE-3006
Changelog: None

No longer blocked by https://github.com/cfengine/core/pull/3526 !

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a pull request! Maybe @olehermanse can review this?

**Comment by karlhto** [comment]
Note that there is no reason to do a jenkins build before CFE-3001 is merged.

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics :) Let's see where we are.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2166/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/2166/

**Comment by olehermanse** [comment]
@karlhto failing on windows:

```
12:31:38   CC       json-yaml.lo
12:31:38   CC       known_dirs.lo
12:31:39   CC       libcrypto-compat.lo
12:31:39 In file included from known_dirs.c:29:0:
12:31:39 ./cf-windows-functions.h:40:40: error: unknown type name 'CfSize'
12:31:39  off_t NovaWin_GetDiskUsage(char *file, CfSize type);
12:31:39                                         ^
12:31:39 ./cf-windows-functions.h:47:38: error: unknown type name 'LogLevel'
12:31:39  void LogToSystemLog(const char *msg, LogLevel level);
12:31:39                                       ^
12:31:39 ./cf-windows-functions.h:58:32: error: unknown type name 'Item'
12:31:39  int NovaWin_GetProcessSnapshot(Item **procdata);
12:31:39                                 ^
12:31:39 ./cf-windows-functions.h:59:24: error: unknown type name 'Item'
12:31:39  int GatherProcessUsers(Item **userList, int *userListSz,
12:31:39                         ^
12:31:39 Makefile:627: recipe for target 'known_dirs.lo' failed
12:31:39 make[3]: *** [known_dirs.lo] Error 1
12:31:39   CC       list.lo
12:31:40   CC       logging.lo
```

**Comment by karlhto** [comment]
Alright, so I need to clean up the commit history here before it's ready.

**Comment by karlhto** [comment]
Started new build for MINGW only:

[Build Status](https://ci.cfengine.com/job/pr-pipeline/2184/)

**Comment by karlhto** [comment]
@olehermanse @vpodzime @craigcomstock This should be ready for review. WIP tag can also be removed.

@cf-bottom Jenkins with exotics.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2210/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2210/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2210/

File: libutils/known_dirs.c [hunk 1]
```diff
@@ -23,11 +23,9 @@
 */
 
 #include <known_dirs.h>
-#include <cf3.defs.h>
+#include <definitions.h>
```
Comment (vpodzime) [inline libutils/known_dirs.c]: Hmm, shouldn't this have been done earlier as part of the commit renaming that file?

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2216/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2216/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2216/

**Comment by olehermanse** [comment]
@karlhto please test the windows packages ^

**Comment by olehermanse** [comment]
@karlhto I tested this for you today. Seems okay;

**Hub:**

```
olehermanse@OHMBP-2840 ~ $ cf-remote info -H 34.242.84.169

ubuntu@34.242.84.169
OS            : ubuntu (debian)
Architecture  : x86_64
CFEngine      : Not installed
Policy server : None
Binaries      : dpkg, apt

olehermanse@OHMBP-2840 ~ $ cf-remote install --hub 34.242.84.169 --hub-package ~/Downloads/cfengine-nova-hub_3.14.0a.88fa30b65_2428_amd64.deb --demo --bootstrap 172.31.37.88

ubuntu@34.242.84.169
OS            : ubuntu (debian)
Architecture  : x86_64
CFEngine      : Not installed
Policy server : None
Binaries      : dpkg, apt

Copying: '/Users/olehermanse/Downloads/cfengine-nova-hub_3.14.0a.88fa30b65_2428_amd64.deb' to '34.242.84.169'
Installing: 'cfengine-nova-hub_3.14.0a.88fa30b65_2428_amd64.deb' on '34.242.84.169'
CFEngine 3.14.0a.88fa30b65 was successfully installed on '34.242.84.169'
Bootstrapping: '34.242.84.169' -> '172.31.37.88'
Bootstrap succesful: '34.242.84.169' -> '172.31.37.88'
Transferring def.json to hub: '34.242.84.169'
Copying: '/Users/olehermanse/.cfengine/cf-remote/json/def.json' to '34.242.84.169'
Triggering an agent run on: '34.242.84.169'
Disabling password change on hub: '34.242.84.169'
Triggering an agent run on: '34.242.84.169'
Your demo hub is ready: https://34.242.84.169/ (Username: admin, Password: password)
olehermanse@OHMBP-2840 ~ $ 
```

*Client:*

```
PS C:\Users\Administrator> .\cfengine-nova-3.14.0a-2431-x86_64.msi                                                                                                                                                                    
PS C:\Users\Administrator> & 'C:\Program Files\Cfengine\bin\cf-agent.exe' -B 172.31.37.88                                                                                                                                             
  notice: Bootstrap mode: implicitly trusting server, use --trust-server=no if server trust is already established                                                                                                                    
  notice: Trusting new key: SHA=2d631840004908f7d12240e0e91877c043e251956202a40095d71d08d73e0601                                                                                                                                      
  notice: Q: "...C:\Program File": R: Should restart cf-serverd because something in its data changed.                                                                                                                                
Q: "...C:\Program File": R: Should restart cf-execd because something in its data changed.                                                                                                                                            
Q: "...C:\Program File": R: Should restart cf-monitord because something in its data changed.                                                                                                                                         
Q: "...C:\Program File": R: Should restart cf-hub because something in its data changed.                                                                                                                                              
R: Bootstrapping from host '172.31.37.88' via built-in policy 'C:\Program Files\Cfengine\inputs\failsafe.cf'                                                                                                                          
R: This autonomous node assumes the role of voluntary client                                                                                                                                                                          
R: Updated local policy from policy server                                                                                                                                                                                            
R: Triggered an initial run of the policy                                                                                                                                                                                             
R: Started the scheduler                                                                                                                                                                                                              
  notice: Bootstrap to '172.31.37.88' completed successfully!                                                                                                                                                                         
PS C:\Users\Administrator> & 'C:\Program Files\Cfengine\bin\cf-agent.exe' -KI                                                                                                                                                         
PS C:\Users\Administrator>  
```

Both ubuntu hub and windows client show up in MP after some minutes. Reporting data is there, Inventory report looks good.

Checked hub syslog for errors after a while:

```
root@ip-172-31-37-88:~# cat /var/log/syslog | grep [c]f-
Mar 26 12:17:15 ip-172-31-37-88 cf-check: CFEngine(generic)  Unrecognized command: '-M'
Mar 26 12:17:25 ip-172-31-37-88 systemd[1]: cf-apache.service: Can't open PID file /var/cfengine/httpd/logs/httpd.pid (yet?) after start: No such file or directory
Mar 26 12:17:28 ip-172-31-37-88 systemd[1]: cf-apache.service: Can't open PID file /var/cfengine/httpd/logs/httpd.pid (yet?) after start: No such file or directory
Mar 26 12:17:28 ip-172-31-37-88 cf-serverd[3321]:   notice: Server is starting...
Mar 26 12:17:28 ip-172-31-37-88 cf-serverd[3321]: CFEngine(server)  Server is starting...
Mar 26 12:17:28 ip-172-31-37-88 cf-agent[3360]: CFEngine(agent)  Storing sha256 hash for '/etc/passwd' (SHA=73bb8988ea3124c6214f2d23101c9b29d419390edd1a20b65c373a864dd50a82)
Mar 26 12:17:29 ip-172-31-37-88 cf-agent[3360]: CFEngine(agent)  Storing sha256 hash for '/etc/group' (SHA=b9c1eba4624a6243174fc843dbde8fd1d2b398cb059ada442a509fd866948333)
Mar 26 12:17:29 ip-172-31-37-88 cf-agent[3360]: CFEngine(agent)  Storing sha256 hash for '/etc/services' (SHA=75d36dfa5c3f829d23f64750d6451641963c5a1e0c6e5d85fb83525755b1450c)
Mar 26 12:17:29 ip-172-31-37-88 cf-agent[3360]: CFEngine(agent)  Storing sha256 hash for '/etc/shadow' (SHA=038b95fd13eb8553b8b0b5d89749fa871a06862b34de5faceafd91589c0a3fe5)
Mar 26 12:17:32 ip-172-31-37-88 systemd[1]: cf-apache.service: Can't open PID file /var/cfengine/httpd/logs/httpd.pid (yet?) after start: No such file or directory
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  Q: ".../cf-agent" --de":   notice: Storing sha256 hash for '/etc/passwd' (SHA=73bb8988ea3124c6214f2d23101c9b29d419390edd1a20b65c373a864dd50a82)#012Q: ".../cf-agent" --de":   notice: Storing sha256 hash for '/etc/group' (SHA=b9c1eba4624a6243174fc843dbde8fd1d2b398cb059ada442a509fd866948333)#012Q: ".../cf-agent" --de":   notice: Storing sha256 hash for '/etc/services' (SHA=75d36dfa5c3f829d23f64750d6451641963c5a1e0c6e5d85fb83525755b1450c)#012Q: ".../cf-agent" --de":   notice: Storing sha256 hash for '/etc/shadow' (SHA=038b95fd13eb8553b8b0b5d89749fa871a06862b34de5faceafd91589c0a3fe5)
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  R: Bootstrapping from host '172.31.37.88' via built-in policy '/var/cfengine/inputs/failsafe.cf'
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  R: This host assumes the role of policy server
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  R: Updated local policy from policy server
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  R: Triggered an initial run of the policy
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  R: Restarted systemd unit cfengine3
Mar 26 12:17:32 ip-172-31-37-88 cf-agent[3280]: CFEngine(agent)  Bootstrap to '172.31.37.88' completed successfully!
Mar 26 12:17:34 ip-172-31-37-88 cf-serverd[3321]:   notice: Cleaning up and exiting...
Mar 26 12:17:34 ip-172-31-37-88 cf-serverd[3321]: CFEngine(server)  Cleaning up and exiting...
Mar 26 12:17:34 ip-172-31-37-88 cf-serverd[3663]:   notice: Server is starting...
Mar 26 12:17:34 ip-172-31-37-88 cf-serverd[3663]: CFEngine(server)  Server is starting...
Mar 26 12:17:38 ip-172-31-37-88 cf-agent[3628]: CFEngine(agent)  R: Should restart cf-serverd because something in its data changed.
Mar 26 12:17:38 ip-172-31-37-88 cf-agent[3628]: CFEngine(agent)  R: Should restart cf-execd because something in its data changed.
Mar 26 12:17:38 ip-172-31-37-88 cf-agent[3628]: CFEngine(agent)  R: Should restart cf-monitord because something in its data changed.
Mar 26 12:17:38 ip-172-31-37-88 cf-agent[3628]: CFEngine(agent)  R: Should restart cf-hub because something in its data changed.
Mar 26 12:17:41 ip-172-31-37-88 systemd[1]: cf-runalerts.service: Service hold-off time over, scheduling restart.
Mar 26 12:17:41 ip-172-31-37-88 systemd[1]: cf-runalerts.service: Scheduled restart job, restart counter is at 1.
Mar 26 12:17:45 ip-172-31-37-88 cf-serverd[3663]:   notice: Rereading policy file '/var/cfengine/inputs/promises.cf'
Mar 26 12:17:45 ip-172-31-37-88 cf-serverd[3663]: CFEngine(server)  Rereading policy file '/var/cfengine/inputs/promises.cf'
Mar 26 12:17:45 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Processing host events
Mar 26 12:17:45 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:17:45 ip-172-31-37-88 cf-serverd[3663]:  warning: 172.31.37.88>    Diff history cleanup. File '/var/cfengine/state/diff/patch.diff' not found.
Mar 26 12:17:45 ip-172-31-37-88 cf-serverd[3663]: CFEngine(server)  172.31.37.88>    Diff history cleanup. File '/var/cfengine/state/diff/patch.diff' not found.
Mar 26 12:17:45 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:22:44 ip-172-31-37-88 cf-hub[3713]: message repeated 12 times: [ NOTICE:  Host bootstrap triggering refreshing m_hosts]
Mar 26 12:23:09 ip-172-31-37-88 cf-serverd[3663]:   notice: 172.31.47.186>   Trusting new key: SHA=f39fb0ecef7606262a9a87f40cd10a115ff31920daa2a9aefb548e64573e4bb3
Mar 26 12:23:09 ip-172-31-37-88 cf-serverd[3663]: CFEngine(server)  172.31.47.186>   Trusting new key: SHA=f39fb0ecef7606262a9a87f40cd10a115ff31920daa2a9aefb548e64573e4bb3
Mar 26 12:23:44 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:23:44 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:24:44 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:24:44 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:25:43 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Processing host events
Mar 26 12:25:43 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:25:45 ip-172-31-37-88 cf-hub[3713]: message repeated 5 times: [ NOTICE:  Host bootstrap triggering refreshing m_hosts]
Mar 26 12:26:43 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
Mar 26 12:34:45 ip-172-31-37-88 cf-hub[3713]: message repeated 37 times: [ NOTICE:  Host bootstrap triggering refreshing m_hosts]
Mar 26 12:35:43 ip-172-31-37-88 cf-hub[3713]: NOTICE:  Host bootstrap triggering refreshing m_hosts
root@ip-172-31-37-88:~# 
```

**Comment by olehermanse** [comment]
@karlhto I resolved your conflict in the GitHub UI. Note that changes like that one (reordering a file or reindenting) is likely to cause a conflict if someone else does some work on that file. To avoid this you could submit it as a separate PR and get it merged as fast as possible.

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2223/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2223/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2223/

**Comment by olehermanse** [comment]
Replaced by: https://github.com/cfengine/core/pull/3569

**Pull request closed without merging**